ocaml/alerts/expiry_alert.ml

```diff
@@ -0,0 +1,112 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module XenAPI = Client.Client
+module Date = Xapi_stdext_date.Date
+
+type message_name_t = string
+
+type message_priority_t = int64
+
+type message_id_t = message_name_t * message_priority_t
+
+type remaining_days_t = int
+
+type raw_alert_t = {
+    cls:
+      [ `Certificate
+      | `Host
+      | `PVS_proxy
+      | `Pool
+      | `SR
+      | `VDI
+      | `VM
+      | `VMPP
+      | `VMSS ]
+  ; obj_uuid: string
+  ; obj_description: string
+  ; alert_conditions: (remaining_days_t * message_id_t) list
+  ; expiry: Xapi_stdext_date.Date.t (* when the obj will expire *)
+}
+
+let seconds_per_day = 3600. *. 24.
+
+let days_until_expiry epoch expiry =
+  (expiry /. seconds_per_day) -. (epoch /. seconds_per_day)
+
+let all_messages rpc session_id =
+  XenAPI.Message.get_all_records ~rpc ~session_id
+
+let message_body msg expiry =
+  Printf.sprintf "<body><message>%s</message><date>%s</date></body>" msg
+    (Date.to_string expiry)
+
+let expired_message obj = Printf.sprintf "The %s has expired." obj
+
+let expiring_message obj = Printf.sprintf "The %s is expiring soon." obj
+
+let generate_alert now obj_description alert_conditions expiry =
+  let remaining_days =
+    days_until_expiry (Date.to_float now) (Date.to_float expiry)
+  in
+  alert_conditions
+  |> List.sort (fun (a, _) (b, _) -> compare a b)
+  |> List.find_opt (fun (days, _) -> remaining_days < float_of_int days)
+  |> function
+  | None ->
+      None
+  | Some (days, expired_message_id) when days <= 0 ->
+      let expired = expired_message obj_description in
+      Some (message_body expired expiry, expired_message_id)
+  | Some (_, expiring_message_id) ->
+      let expiring = expiring_message obj_description in
+      Some (message_body expiring expiry, expiring_message_id)
+
+let filter_messages msg_name_list msg_obj_uuid alert all_msgs =
```

Comment:
This is undocumented both here and in the MLI.

Reply:
Added comment for this function.